Conditional dimension depending on muliple dimensions placed in innermost loop

Consider this proxy for a Fourier integral:
```
from devito import (TimeFunction, ConditionalDimension, Function,
                    DefaultDimension, Grid, Eq, Operator)
from sympy import Ge
import numpy as np

freq = [2,4,8]
nfreq = np.shape(freq)[0]
grid = Grid(shape=(5,5))
# time function for holding wavefield proxy
u = TimeFunction(name="u", grid=grid)
t = grid.time_dim
# frequency dimension for holding fft proxy
freq_dim = DefaultDimension(name='freq_dim', default_value=nfreq)
f = Function(name="f", grid=grid, dimensions=(freq_dim,), shape=(nfreq,))
# load the frequencies into f, this will serve as a lower bound for the proxy integral
f.data[:] = np.array(freq[:])
# proxy fourier integral holder
ure = Function(name="ure", grid=grid, dimensions=(freq_dim,) + u.indices[1:], shape=(nfreq,) + u.shape[1:])
# make update for u
u_up = Eq(u.forward, u+1)
# conditional dimension based on f to simulate bounds of fourier integral
ct = ConditionalDimension(name="ct", parent=t, condition=Ge(t,f), indirect=True)
# proxy for fourier integral
fourier_eq = Eq(ure, ure + u, implicit_dims=ct)
op = Operator([u_up, fourier_eq])
op.apply(time_M=10)

print(ure.data)

```

It returns the following correct result:
```
[[[54. 54. 54. 54. 54.]
  [54. 54. 54. 54. 54.]
  [54. 54. 54. 54. 54.]
  [54. 54. 54. 54. 54.]
  [54. 54. 54. 54. 54.]]

 [[49. 49. 49. 49. 49.]
  [49. 49. 49. 49. 49.]
  [49. 49. 49. 49. 49.]
  [49. 49. 49. 49. 49.]
  [49. 49. 49. 49. 49.]]

 [[27. 27. 27. 27. 27.]
  [27. 27. 27. 27. 27.]
  [27. 27. 27. 27. 27.]
  [27. 27. 27. 27. 27.]
  [27. 27. 27. 27. 27.]]]
```

However, if we examine the C code we find that the conditional dimension acting as the integral bound has been placed within the innermost loop:
```
#define _POSIX_C_SOURCE 200809L
#define MIN(a,b) (((a) < (b)) ? (a) : (b))
#define MAX(a,b) (((a) > (b)) ? (a) : (b))
#define START_TIMER(S) struct timeval start_ ## S , end_ ## S ; gettimeofday(&start_ ## S , NULL);
#define STOP_TIMER(S,T) gettimeofday(&end_ ## S, NULL); T->S += (double)(end_ ## S .tv_sec-start_ ## S.tv_sec)+(double)(end_ ## S .tv_usec-start_ ## S .tv_usec)/1000000;

#include "stdlib.h"
#include "math.h"
#include "sys/time.h"
#include "xmmintrin.h"
#include "pmmintrin.h"

struct dataobj
{
  void *restrict data;
  unsigned long * size;
  unsigned long * npsize;
  unsigned long * dsize;
  int * hsize;
  int * hofs;
  int * oofs;
  void * dmap;
} ;

struct profiler
{
  double section0;
  double section1;
} ;


int Kernel(struct dataobj *restrict f_vec, struct dataobj *restrict u_vec, struct dataobj *restrict ure_vec, const int freq_dim0_blk0_size, const int freq_dim_M, const int freq_dim_m, const int time_M, const int time_m, const int x0_blk0_size, const int x_M, const int x_m, const int y_M, const int y_m, struct profiler * timers)
{
  float (*restrict f) __attribute__ ((aligned (64))) = (float (*)) f_vec->data;
  float (*restrict u)[u_vec->size[1]][u_vec->size[2]] __attribute__ ((aligned (64))) = (float (*)[u_vec->size[1]][u_vec->size[2]]) u_vec->data;
  float (*restrict ure)[ure_vec->size[1]][ure_vec->size[2]] __attribute__ ((aligned (64))) = (float (*)[ure_vec->size[1]][ure_vec->size[2]]) ure_vec->data;

  /* Flush denormal numbers to zero in hardware */
  _MM_SET_DENORMALS_ZERO_MODE(_MM_DENORMALS_ZERO_ON);
  _MM_SET_FLUSH_ZERO_MODE(_MM_FLUSH_ZERO_ON);

  for (int time = time_m, t0 = (time)%(2), t1 = (time + 1)%(2); time <= time_M; time += 1, t0 = (time)%(2), t1 = (time + 1)%(2))
  {
    /* Begin section0 */
    START_TIMER(section0)
    for (int x = x_m; x <= x_M; x += 1)
    {
      #pragma omp simd aligned(u:64)
      for (int y = y_m; y <= y_M; y += 1)
      {
        u[t1][x + 1][y + 1] = u[t0][x + 1][y + 1] + 1;
      }
    }
    STOP_TIMER(section0,timers)
    /* End section0 */

    /* Begin section1 */
    START_TIMER(section1)
    for (int freq_dim0_blk0 = freq_dim_m; freq_dim0_blk0 <= freq_dim_M; freq_dim0_blk0 += freq_dim0_blk0_size)
    {
      for (int x0_blk0 = x_m; x0_blk0 <= x_M; x0_blk0 += x0_blk0_size)
      {
        for (int freq_dim = freq_dim0_blk0; freq_dim <= MIN(freq_dim0_blk0 + freq_dim0_blk0_size - 1, freq_dim_M); freq_dim += 1)
        {
          for (int x = x0_blk0; x <= MIN(x0_blk0 + x0_blk0_size - 1, x_M); x += 1)
          {
            #pragma omp simd aligned(f,u,ure:64)
            for (int y = y_m; y <= y_M; y += 1)
            {
              if (time >= f[freq_dim])
              {
                ure[freq_dim][x + 1][y + 1] = u[t0][x + 1][y + 1] + ure[freq_dim][x + 1][y + 1];
              }
            }
          }
        }
      }
    }
    STOP_TIMER(section1,timers)
    /* End section1 */
  }

  return 0;
}
```

If I modify the conditional dimension definition to be the following:
`ct = ConditionalDimension(name="ct", parent=t, condition=Ge(t,f), indirect=False)`
I get the following compiler error:
```
gcc -O3 -g -fPIC -Wall -std=c99 -march=native -Wno-unused-result -Wno-unused-variable -Wno-unused-but-set-variable -ffast-math -shared -fopenmp /tmp/devito-jitcache-uid1000/bb75d42a661d62a79aa80ab0ea52cddc47b82dff.c -lm -o /tmp/devito-jitcache-uid1000/bb75d42a661d62a79aa80ab0ea52cddc47b82dff.so
/tmp/devito-jitcache-uid1000/bb75d42a661d62a79aa80ab0ea52cddc47b82dff.c: In function ‘Kernel’:
/tmp/devito-jitcache-uid1000/bb75d42a661d62a79aa80ab0ea52cddc47b82dff.c:56:19: error: ‘freq_dim’ undeclared (first use in this function); did you mean ‘freq_dim_m’?
   56 |     if (time >= f[freq_dim])
      |                   ^~~~~~~~
      |                   freq_dim_m
```

If I modify the conditional dimension and the implicit dimensions of the integral equation like this:
`ct = ConditionalDimension(name="ct", parent=t, condition=Ge(t,f), indirect=False)`
`fourier_eq = Eq(ure, ure + u, implicit_dims=(freq_dim, ct))`

It compiles but I get the following incorrect result:
```
[[[94. 94. 94. 94. 94.]
  [94. 94. 94. 94. 94.]
  [94. 94. 94. 94. 94.]
  [94. 94. 94. 94. 94.]
  [94. 94. 94. 94. 94.]]

 [[73. 73. 73. 73. 73.]
  [73. 73. 73. 73. 73.]
  [73. 73. 73. 73. 73.]
  [73. 73. 73. 73. 73.]
  [73. 73. 73. 73. 73.]]

 [[31. 31. 31. 31. 31.]
  [31. 31. 31. 31. 31.]
  [31. 31. 31. 31. 31.]
  [31. 31. 31. 31. 31.]
  [31. 31. 31. 31. 31.]]]
```

Examining the code, this is because a second time loop has been created following the updates to U, so the integral is only really updated based on U at the last time step:
```
#define _POSIX_C_SOURCE 200809L
#define START_TIMER(S) struct timeval start_ ## S , end_ ## S ; gettimeofday(&start_ ## S , NULL);
#define STOP_TIMER(S,T) gettimeofday(&end_ ## S, NULL); T->S += (double)(end_ ## S .tv_sec-start_ ## S.tv_sec)+(double)(end_ ## S .tv_usec-start_ ## S .tv_usec)/1000000;

#include "stdlib.h"
#include "math.h"
#include "sys/time.h"
#include "xmmintrin.h"
#include "pmmintrin.h"

struct dataobj
{
  void *restrict data;
  unsigned long * size;
  unsigned long * npsize;
  unsigned long * dsize;
  int * hsize;
  int * hofs;
  int * oofs;
  void * dmap;
} ;

struct profiler
{
  double section0;
  double section1;
} ;


int Kernel(struct dataobj *restrict f_vec, struct dataobj *restrict u_vec, struct dataobj *restrict ure_vec, const int freq_dim_M, const int freq_dim_m, const int time_M, const int time_m, const int x_M, const int x_m, const int y_M, const int y_m, struct profiler * timers)
{
  float (*restrict f) __attribute__ ((aligned (64))) = (float (*)) f_vec->data;
  float (*restrict u)[u_vec->size[1]][u_vec->size[2]] __attribute__ ((aligned (64))) = (float (*)[u_vec->size[1]][u_vec->size[2]]) u_vec->data;
  float (*restrict ure)[ure_vec->size[1]][ure_vec->size[2]] __attribute__ ((aligned (64))) = (float (*)[ure_vec->size[1]][ure_vec->size[2]]) ure_vec->data;

  /* Flush denormal numbers to zero in hardware */
  _MM_SET_DENORMALS_ZERO_MODE(_MM_DENORMALS_ZERO_ON);
  _MM_SET_FLUSH_ZERO_MODE(_MM_FLUSH_ZERO_ON);

  for (int time = time_m, t0 = (time)%(2), t1 = (time + 1)%(2); time <= time_M; time += 1, t0 = (time)%(2), t1 = (time + 1)%(2))
  {
    /* Begin section0 */
    START_TIMER(section0)
    for (int x = x_m; x <= x_M; x += 1)
    {
      #pragma omp simd aligned(u:64)
      for (int y = y_m; y <= y_M; y += 1)
      {
        u[t1][x + 1][y + 1] = u[t0][x + 1][y + 1] + 1;
      }
    }
    STOP_TIMER(section0,timers)
    /* End section0 */
  }

  /* Begin section1 */
  START_TIMER(section1)
  for (int freq_dim = freq_dim_m; freq_dim <= freq_dim_M; freq_dim += 1)
  {
    for (int time = time_m, t0 = (time)%(2); time <= time_M; time += 1, t0 = (time)%(2))
    {
      if (time >= f[freq_dim])
      {
        for (int x = x_m; x <= x_M; x += 1)
        {
          #pragma omp simd aligned(u,ure:64)
          for (int y = y_m; y <= y_M; y += 1)
          {
            ure[freq_dim][x + 1][y + 1] = u[t0][x + 1][y + 1] + ure[freq_dim][x + 1][y + 1];
          }
        }
      }
    }
  }
  STOP_TIMER(section1,timers)
  /* End section1 */

  return 0;
}
```

--- a/devito/ir/clusters/algorithms.py
+++ b/devito/ir/clusters/algorithms.py
@@ -6,7 +6,8 @@ import numpy as np
 import sympy
 
 from devito.exceptions import InvalidOperator
-from devito.ir.support import Any, Backward, Forward, IterationSpace, PARALLEL_IF_ATOMIC
+from devito.ir.support import (Any, Backward, Forward, IterationSpace,
+                               PARALLEL_IF_ATOMIC, pull_dims)
 from devito.ir.clusters.analysis import analyze
 from devito.ir.clusters.cluster import Cluster, ClusterGroup
 from devito.ir.clusters.visitors import Queue, QueueStateful, cluster_pass
@@ -198,8 +199,17 @@ def guard(clusters):
                 else:
                     d = cd.parent
 
+                # If `cd` uses, as condition, an arbitrary SymPy expression, then
+                # we must ensure to nest it inside the last of the Dimensions
+                # appearing in `cd.condition`
+                if cd._factor is not None:
+                    k = d
+                else:
+                    dims = pull_dims(cd.condition)
+                    k = max(dims, default=d, key=lambda i: c.ispace.index(i))
+
                 # Pull `cd` from any expr
-                condition = guards.setdefault(d, [])
+                condition = guards.setdefault(k, [])
                 for e in exprs:
                     try:
                         condition.append(e.conditionals[cd])

--- a/devito/ir/support/utils.py
+++ b/devito/ir/support/utils.py
@@ -4,7 +4,7 @@ from devito.symbolics import CallFromPointer, retrieve_indexed, retrieve_termina
 from devito.tools import DefaultOrderedDict, as_tuple, flatten, filter_sorted, split
 from devito.types import Dimension, Indirection, ModuloDimension
 
-__all__ = ['AccessMode', 'Stencil', 'detect_accesses', 'detect_io']
+__all__ = ['AccessMode', 'Stencil', 'detect_accesses', 'detect_io', 'pull_dims']
 
 
 class AccessMode(object):
@@ -241,3 +241,18 @@ def detect_io(exprs, relax=False):
                 writes.append(f)
 
     return filter_sorted(reads), filter_sorted(writes)
+
+
+def pull_dims(exprs, flag=True):
+    """
+    Extract all Dimensions from one or more expressions. If `flag=True`
+    (default), all of the ancestor and descendant Dimensions are extracted
+    as well.
+    """
+    dims = set()
+    for e in as_tuple(exprs):
+        dims.update({i for i in e.free_symbols if i.is_Dimension})
+    if flag:
+        return set().union(*[d._defines for d in dims])
+    else:
+        return dims

--- a/devito/passes/clusters/blocking.py
+++ b/devito/passes/clusters/blocking.py
@@ -184,6 +184,10 @@ class AnalyzeHeuristicBlocking(AnayzeBlockingBase):
             return clusters
 
         d = prefix[-1].dim
+        if d.is_Default:
+            # Heuristic: DefaultDimensions typically define relatively small
+            # iteration spaces, hence they're ruled out
+            return clusters
 
         for c in clusters:
             # PARALLEL* and AFFINE are necessary conditions

--- a/tests/test_dimension.py
+++ b/tests/test_dimension.py
@@ -7,8 +7,8 @@ import pytest
 from conftest import assert_blocking, skipif, opts_tiling
 from devito import (ConditionalDimension, Grid, Function, TimeFunction,  # noqa
                     SparseFunction, SparseTimeFunction, Eq, Operator, Constant,
-                    Dimension, SubDimension, switchconfig, SubDomain, Lt, Le,
-                    Gt, Ge, Ne, Buffer, sin)
+                    Dimension, DefaultDimension, SubDimension, switchconfig,
+                    SubDomain, Lt, Le, Gt, Ge, Ne, Buffer, sin)
 from devito.ir.iet import (Conditional, Expression, Iteration, FindNodes,
                            retrieve_iteration_tree)
 from devito.symbolics import indexify, retrieve_functions, IntDiv
@@ -1213,6 +1213,50 @@ class TestConditionalDimension(object):
         assert all(p.data[i].sum() == i - 1 for i in range(1, 12))
         assert all(p.data[i, 10, 10, 10] == i - 1 for i in range(1, 12))
         assert all(np.all(p.data[i] == 0) for i in range(12, 20))
+
+    def test_issue_2007(self):
+        """
+        Proxy for a Fourier integral. Main issue: conditional placed too deep
+        in the tree. It's functionally OK, but performance-wise really bad.
+        """
+        freq = [2, 4, 8]
+        nfreq = np.shape(freq)[0]
+        grid = Grid(shape=(5, 5))
+        time = grid.time_dim
+
+        u = TimeFunction(name="u", grid=grid)
+
+        # frequency dimension for holding fft proxy
+        freq_dim = DefaultDimension(name='freq_dim', default_value=nfreq)
+        f = Function(name="f", grid=grid, dimensions=(freq_dim,), shape=(nfreq,))
+
+        # Load the frequencies into f, this will serve as a lower bound for the
+        # proxy integral
+        f.data[:] = np.array(freq[:])
+        # Proxy Fourier integral holder
+        ure = Function(name="ure", grid=grid,
+                       dimensions=(freq_dim,) + u.indices[1:],
+                       shape=(nfreq,) + u.shape[1:])
+
+        # ConditionalDimension based on `f` to simulate bounds of Fourier integral
+        ct = ConditionalDimension(name="ct", parent=time, condition=Ge(time, f))
+        eqns = [
+            Eq(u.forward, u+1),
+            Eq(ure, ure + u, implicit_dims=ct)
+        ]
+
+        op = Operator(eqns)
+
+        op.apply(time_M=10)
+
+        assert np.all(ure.data[0] == 54)
+        assert np.all(ure.data[1] == 49)
+        assert np.all(ure.data[2] == 27)
+
+        # Make sure the ConditionalDimension is at the right depth for performance
+        trees = retrieve_iteration_tree(op)
+        assert len(trees) == 3
+        assert trees[1][1].nodes[0].is_Conditional
 
 
 class TestMashup(object):

--- a/tests/test_dse.py
+++ b/tests/test_dse.py
@@ -1419,8 +1419,7 @@ class TestAliases(object):
         op = Operator([pde, df])
 
         # Check code generation
-        assert_structure(op, ['t,x,y', 't,f', 't,f0_blk0,x0_blk0,f,x,y'],
-                         't,x,y,f,f0_blk0,x0_blk0,f,x,y')
+        assert_structure(op, ['t,x,y', 't,f', 't,f,x,y'], 't,x,y,f,f,x,y')
 
     def test_space_invariant_v2(self):
         """

